### PR TITLE
Add OpenTelemetry tracing

### PR DESCRIPTION
**Pull request by electrum**
## Description

Add initial, experimental support for distributed tracing using [OpenTelemetry](https://opentelemetry.io/). This adds [traces](https://opentelemetry.io/docs/concepts/signals/traces/) for queries with [spans](https://opentelemetry.io/docs/concepts/signals/traces/#spans) for analysis, planning, optimization, split generation, tasks, remote tasks, local planning, metadata, HTTP client requests, JAX-RS endpoints, etc.

Connectors can participate in tracing using the OpenTelemetry API. As a proof-of-concept, `TrinoFileSystem` and the `S3` file system code are instrumented for the Hive, Iceberg, Delta Lake and Hudi connectors.

## Usage

Enable tracing in Trino and configure the [OLTP/gRPC endpoint](https://opentelemetry.io/docs/reference/specification/protocol/otlp/):

```properties
tracing.enabled=true
tracing.exporter.endpoint=http://localhost:4317
```

For easy testing on a single machine, such as when running `IcebergMinioQueryRunnerMain`, you can run [Jaeger](https://www.jaegertracing.io/docs/latest/getting-started/) in Docker:

```bash
docker run -e COLLECTOR_OTLP_ENABLED=true -p 16686:16686 -p 4317:4317 jaegertracing/all-in-one:latest
```

Open `http://localhost:16686` in your browser to view the Jaeger UI.

## Additional context and related issues

* Depends on https://github.com/airlift/airlift/pull/1055
* Depends on https://github.com/airlift/discovery/pull/49
* Fixes https://github.com/trinodb/trino/issues/1826
* Fixes https://github.com/trinodb/trino/issues/4207

## Release notes

(x) Release notes are required, with the following suggested text:

```markdown
# General
* Add experimental support for tracing using [OpenTelemetry](https://opentelemetry.io/).
  Enable tracing by setting `tracing.enabled=true` and configure the 
  [OLTP/gRPC endpoint](https://opentelemetry.io/docs/reference/specification/protocol/otlp/)
  endpoint by setting `tracing.exporter.endpoint` to override
  the default of `http://localhost:4317` if needed. The functionality, span names, attributes, etc.,
  are experimental and subject to change. ({issue}`16950`)
```

### CODE REVIEW

**Comment by sajjoseph** [comment]
Long overdue feature. Thanks @electrum for making it happen.

File: core/trino-main/src/main/java/io/trino/execution/QueryStateMachine.java [hunk 1]
```diff
@@ -320,6 +336,24 @@ static QueryStateMachine beginWithTicker(
             }
         });
 
+        queryStateMachine.addStateChangeListener(newState -> {
+            querySpan.addEvent("query_state", Attributes.of(
+                    TrinoAttributes.EVENT_STATE, newState.toString()));
```
Comment (dain) [inline core/trino-main/src/main/java/io/trino/execution/QueryStateMachine.java:341]: Did you mean to wrap here?

File: core/trino-main/src/main/java/io/trino/execution/SqlTaskExecutionFactory.java [hunk 2]
```diff
@@ -73,7 +80,7 @@ public SqlTaskExecution create(
 
         LocalExecutionPlan localExecutionPlan;
         try (SetThreadName ignored = new SetThreadName("Task-%s", taskStateMachine.getTaskId())) {
-            try {
+            try (var ignored2 = scopedSpan(tracer, "local-planner")) {
```
Comment (dain) [inline core/trino-main/src/main/java/io/trino/execution/SqlTaskExecutionFactory.java:83]: did you mean `ignored2`?

File: core/trino-main/src/main/java/io/trino/sql/planner/LogicalPlanner.java [hunk 3]
```diff
@@ -280,11 +283,54 @@ public Plan plan(Analysis analysis, Stage stage, boolean collectPlanStatistics)
...
 
+    @Nonnull
+    private PlanNode runOptimizer(PlanNode root, TableStatsProvider tableStatsProvider, PlanOptimizer optimizer)
+    {
+        PlanNode result;
+        try (var ignored = optimizerSpan(optimizer)) {
+            result = optimizer.optimize(root, session, symbolAllocator.getTypes(), symbolAllocator, idAllocator, warningCollector, planOptimizersStatsCollector, tableStatsProvider);
+        }
+        if (root == null) {
+            throw new NullPointerException(optimizer.getClass().getName() + " returned a null plan");
```
Comment (dain) [inline core/trino-main/src/main/java/io/trino/sql/planner/LogicalPlanner.java:301]: Should be mark the span as an error in this case?

File: core/trino-main/src/main/java/io/trino/tracing/TracingMetadata.java [hunk 4]
```diff
@@ -0,0 +1,1393 @@
...
+        }
+        return span;
+    }
+
+    private Span startSpan(String methodName, TableExecuteHandle handle)
+    {
+        Span span = startSpan(methodName);
+        if (span.isRecording()) {
+            span.setAttribute(TrinoAttributes.CATALOG, handle.getCatalogHandle().getCatalogName());
+            span.setAttribute(TrinoAttributes.HANDLE, handle.getConnectorHandle().toString());
```
Comment (dain) [inline core/trino-main/src/main/java/io/trino/tracing/TracingMetadata.java:1380]: is the handle going to be too long in text form or possibly leak push down sql?

File: core/trino-main/src/main/java/io/trino/tracing/TracingConnectorMetadata.java [hunk 5]
```diff
@@ -0,0 +1,1280 @@
...
+        return startSpan(methodName)
+                .setAllAttributes(attribute(TrinoAttributes.SCHEMA, prefix.getSchema()))
+                .setAllAttributes(attribute(TrinoAttributes.TABLE, prefix.getTable()));
+    }
+
+    private Span startSpan(String methodName, ConnectorTableHandle handle)
+    {
+        Span span = startSpan(methodName);
+        if (span.isRecording()) {
+            span.setAttribute(TrinoAttributes.HANDLE, handle.toString());
```
Comment (dain) [inline core/trino-main/src/main/java/io/trino/tracing/TracingConnectorMetadata.java:1258]: same question about `handle.toString`

File: lib/trino-filesystem/src/main/java/io/trino/filesystem/tracing/TracingInput.java [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+        this.location = requireNonNull(location, "location is null");
+        this.fileLength = requireNonNull(fileLength, "fileLength is null");
+    }
+
+    @Override
+    public void readFully(long position, byte[] buffer, int bufferOffset, int bufferLength)
+            throws IOException
+    {
+        Span span = spanBuilder("Input.readFully", bufferLength)
+                .setAttribute(FileSystemAttributes.FILE_READ_POSITION, position)
```
Comment (dain) [inline lib/trino-filesystem/src/main/java/io/trino/filesystem/tracing/TracingInput.java:50]: maybe add read size?

**Review by dain — APPROVED** [review]

File: core/trino-main/src/main/java/io/trino/tracing/TracingConnectorMetadata.java [hunk 7]
```diff
@@ -0,0 +1,1280 @@
...
+import java.util.OptionalInt;
+import java.util.OptionalLong;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+
+import static io.airlift.tracing.Tracing.attribute;
+import static io.trino.tracing.ScopedSpan.scopedSpan;
+import static java.util.Objects.requireNonNull;
+
+public class TracingConnectorMetadata
```
Comment (wendigo) [inline core/trino-main/src/main/java/io/trino/tracing/TracingConnectorMetadata.java:97]: Are these classed autogenerated ?

**Comment by wendigo** [comment]
Is there a kill switch to disable tracing? What's the added overhead?

**Comment by electrum** [comment]
@wendigo It is only enabled if you do `tracing.enabled`

File: core/trino-main/src/main/java/io/trino/tracing/TracingConnectorMetadata.java (hunk 7 above)
Comment (electrum) [reply core/trino-main/src/main/java/io/trino/tracing/TracingConnectorMetadata.java:97]: Nope, I wrote them by hand.

File: core/trino-main/src/main/java/io/trino/tracing/TracingMetadata.java (hunk 4 above)
Comment (electrum) [reply core/trino-main/src/main/java/io/trino/tracing/TracingMetadata.java:1379]: We print these in plans, so it's probably fine

File: lib/trino-filesystem/src/main/java/io/trino/filesystem/tracing/TracingInput.java (hunk 6 above)
Comment (electrum) [reply lib/trino-filesystem/src/main/java/io/trino/filesystem/tracing/TracingInput.java:50]: We log that above using `spanBuilder`

**Comment by bitsondatadev** [comment]
This is incredible!!!!

**Comment by colebow** [comment]
@electrum - do we want to document this, do you have that covered, or should we wait until it's not experimental?

**Comment by lambrospetrou** [comment]
I looked through the code, but cannot find anything.
Is there a way we can specify the `tracing.exporter.endpoint` override using an environment variable instead of the config file?
In our case, we use k8s, and we need to dynamically retrieve the host IP of the OTLP exporter endpoint and then pass it to Trino as environment variable.

**Pull request merged**